### PR TITLE
Recent improvements (including support for UC Berkely spice)

### PR DESCRIPTION
**Pull request by gtrentalancia**
Please merge, several improvements have been made.

### CODE REVIEW

**Comment by gtrentalancia** [comment]
Also, after merging, please create a new release.

Least but not last, I don't know how to fix the following issue:

https://github.com/drahnr/oregano/issues/199

can you help ?

Thanks !

File: src/engines/ngspice-analysis.c [hunk 1]
```diff
@@ -861,7 +861,7 @@ static ThreadPipe *parse_noise_analysis (NgspiceAnalysisResources *resources)
         static SimulationData *sdata;
         static Analysis *data;
 	FILE *fp;
-	char fbuffer[BSIZE_SP];
+	gchar *fbuffer;
```
Comment (drahnr) [inline src/engines/ngspice-analysis.c]: why? what is the advantage?

File: src/engines/ngspice-analysis.c [hunk 2]
```diff
@@ -933,15 +937,15 @@ static ThreadPipe *parse_noise_analysis (NgspiceAnalysisResources *resources)
...
+							fbuffer, 0, 0);
 		if (splitted[1]) {
 			if (*splitted[1]) {
 				prev_seq = seq;
 				seq = g_ascii_strtod(splitted[1], NULL);
 
 				// simple sequence sanity check
 				if (seq != prev_seq + 1)
-					return pipe;
+					break;
```
Comment (drahnr) [inline src/engines/ngspice-analysis.c]: use `g_autofree` for `fbuffer` in case it really has to be allocated dynamically

File: src/settings.c [hunk 3]
```diff
@@ -131,17 +131,17 @@ void settings_show (GtkWidget *widget, SchematicView *sv)
...
 	g_return_if_fail (sv != NULL);
 
-	// If no engine available, do not show the preferences
-	if ((g_find_program_in_path (engine[0]) == NULL) &&
-	    (g_find_program_in_path (engine[1]) == NULL) &&
-	     g_find_program_in_path (engine[2]) == NULL) {
+	for (i = 0; i < OREGANO_ENGINE_COUNT; i++)
+		if (g_find_program_in_path(engine[i]) != NULL)
+			engine_available = TRUE;
+
```
Comment (drahnr) [inline src/settings.c]: Please add brackets

**Comment by drahnr** [comment]
I don't see a point in making more releases manually other than tagging.

The CI is not yet capable of doing that, mostly because I need to find an S3 compat storage for cheap, considering minio atm but nothing will happen within the next two weeks.

**Comment by gtrentalancia** [comment]
Supporting the original spice version 3 from UC Berkeley is more important than supporting ngspice, because the original spice is more widely used than ngspice.

We should definitely need to have such feature for a complete application that suits the needs of all users.

Consider that there is no additional maintenance to carry out, because the two engines behave almost the same (ngspice is derived from spice3, so if spice3 changes in some way, ngspice will likely change accordingly).

The only differences between the two engines are as follows:

- spice3 does not print the number of data rows for each kind of analysis: however at the moment such information is not used, so there is no problem;
- during AC Analysis with spice3, the second column (after the frequency for the x-axis) has to be skipped.

It is unlikely that the output format will change for spice3 or ngspice. In any case, in the unlikely event that it changes, I will modify the oregano code accordingly: it is a very easy operation because it is just a matter of parsing a slightly different output from the engine.

Please merge the most recent code from my repository to get all the latest updates, improvements and new features !

Thanks,

Guido

**Comment by drahnr** [comment]
Again, why did you use the HEAP instead of the STACK as of my review comment?

**Comment by gtrentalancia** [comment]
The code that was using the fbuffer variable is no longer there, it has been removed ! Please update your copy of my repository !

**Comment by gtrentalancia** [comment]
I knew that such code was going to be disposed shortly in the next commit, so I didn't even bother to look at how variables were being allocated.

However, the difference is that the code using the heap is just a bit slower, there is no other difference. With such function being called only once every simulation run, the difference is unnoticeable.

Do not consider that obsolete code, please merge the latest one !!

Regards,

Guido

**Comment by drahnr** [comment]
Squash those commits then.

**Comment by gtrentalancia** [comment]
I don't really understand what the problem is...

What is wrong with the code that I have created ? It looks fine to me and it seems to be working fine, so where is the problem ??

**Comment by gtrentalancia** [comment]
But thanks for merging it, you've done the right thing, absolutely !

**Comment by drahnr** [comment]
It is not about the resulting code, it is about the fact that I want them in atomic, topological chunks. It does not matter how you came to your solution, that does not yield any gain, but there is a gain if every changeset is atomic.

**Comment by gtrentalancia** [comment]
What is ${TREEISH} ?

**Comment by drahnr** [comment]
You should really read some git --help pages.

**Comment by gtrentalancia** [comment]
Apart from the git commands, which commit do you want to merge together ?

**Pull request merged**